crates/bevy_pbr/src/render/parallax_mapping.wgsl

```diff
@@ -0,0 +1,102 @@
+#define_import_path bevy_pbr::parallax_mapping
+
+fn sample_depth_map(uv: vec2<f32>) -> f32 {
+    return textureSample(depth_map_texture, depth_map_sampler, uv).r;
+}
+
+// An implementation of parallax mapping, see https://en.wikipedia.org/wiki/Parallax_mapping
+// Code derived from: https://web.archive.org/web/20150419215321/http://sunandblackcat.com/tipFullView.php?l=eng&topicid=28
+fn parallaxed_uv(
+    depth: f32,
+    max_layer_count: f32,
+    // The original uv
+    uv: vec2<f32>,
+    // The vector from camera to the surface of material
+    V: vec3<f32>,
+) -> vec2<f32> {
+    var uv = uv;
+    if max_layer_count < 1.0 {
+        return uv;
+    }
+
+    // Steep parallax mapping
+    // ======================
+    // split the depth map into `layer_count` layers,
+    // When V hits the surface of object (excluding displacement),
+    // if not bellow or on surface including displacement (textureSample), then
+    // look forward (-= delta_uv) according to V and distance between hit surface and
+    // depth map surface, repeat until bellow surface.
+    //
+    // where `layer_count` is selected smartly between `min_layer_count` and
+    // `max_layer_count` according to the steepness of V.
+
+    let view_steepness = abs(dot(vec3<f32>(0.0, 0.0, 1.0), V));
+    // We mix with minimum value 1.0 because otherwise, with 0.0, we get
+    // a nice division by zero in surfaces parallel to viewport, resulting
+    // in a singularity.
+    let layer_count = mix(max_layer_count, 1.0, view_steepness);
+    let layer_height = 1.0 / layer_count;
+    var delta_uv = depth * V.xy / V.z / layer_count;
+
+    var current_layer_height = 0.0;
+    var current_height = sample_depth_map(uv);
+
+    // This at most runs layer_count times
+    while true {
+        if (current_height <= current_layer_height) {
+            break;
+        }
+        current_layer_height += layer_height;
+        uv -= delta_uv;
+        current_height = sample_depth_map(uv);
+    }
```

Comment:
I've hesitated between the `for` and `while` a lot. I chose the `while` because a `for` introduces an additional variable (might increase register stress?) and that `i32(layer_count)` conversion as well. I'll swap back to the `for` loop though